src/mongoc/mongoc-topology.c

```diff
@@ -351,29 +359,23 @@ mongoc_topology_select (mongoc_topology_t         *topology,
          topology->stale = false;
       }
 
-      /* until we find a server or timeout */
-      for (;;) {
-         /* error if we've timed out */
-         now = bson_get_monotonic_time();
-         if (now >= expire_at) {
-            goto TIMEOUT;
-         }
+      /* attempt to select a server */
+      selected_server = mongoc_topology_description_select(&topology->description,
+                                                           optype,
+                                                           read_prefs,
+                                                           local_threshold_ms,
+                                                           error);
 
-         /* attempt to select a server */
-         selected_server = mongoc_topology_description_select(&topology->description,
-                                                              optype,
-                                                              read_prefs,
-                                                              local_threshold_ms,
-                                                              error);
+      if (selected_server) {
+         RETURN (mongoc_server_description_new_copy(selected_server));
+      }
 
-         if (selected_server) {
-            return mongoc_server_description_new_copy(selected_server);
-         }
+      bson_set_error(error,
+                     MONGOC_ERROR_SERVER_SELECTION,
+                     MONGOC_ERROR_SERVER_SELECTION_NO_MATCH,
+                     "Could not find any server matching");
```

Comment:
Yeah, the optype + read_pref should definitely be part of the fallthrough error message